modules/core/02-client/migrations/v7/store.go

```diff
@@ -0,0 +1,215 @@
+// The migration includes:
+//
+// - Migrating solo machine client states from v2 to v3 protobuf definition
+// - Pruning all solo machine consensus states
+// - Removing the localhost client
+// - Asserting existing tendermint clients are properly registered on the chain codec
+func MigrateStore(ctx sdk.Context, storeKey storetypes.StoreKey, cdc codec.BinaryCodec) error {
+	store := ctx.KVStore(storeKey)
+
+	if err := handleSolomachineMigration(ctx, store, cdc); err != nil {
+		return err
+	}
+
+	if err := handleTendermintMigration(ctx, store, cdc); err != nil {
+		return err
+	}
+
+	if err := handleLocalhostMigration(ctx, store, cdc); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// handleSolomachineMigration iterates over the solo machine clients and migrates client state from
+// protobuf definition v2 to v3. All consensus states stored outside of the client state are pruned.
+func handleSolomachineMigration(ctx sdk.Context, store sdk.KVStore, cdc codec.BinaryCodec) error {
+	clients, err := collectClients(ctx, store, exported.Solomachine)
+	if err != nil {
+		return err
+	}
+
+	for _, clientID := range clients {
+		clientPrefix := []byte(fmt.Sprintf("%s/%s/", host.KeyClientStorePrefix, clientID))
+		clientStore := prefix.NewStore(store, clientPrefix)
+
+		bz := clientStore.Get(host.ClientStateKey())
+		if bz == nil {
+			return clienttypes.ErrClientNotFound
+		}
+
+		any := &codectypes.Any{}
+		if err := cdc.Unmarshal(bz, any); err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+		}
+
+		clientState := &ClientState{}
+		if err := cdc.Unmarshal(any.Value, clientState); err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+		}
+
+		updatedClientState := migrateSolomachine(clientState)
+
+		bz, err := clienttypes.MarshalClientState(cdc, updatedClientState)
+		if err != nil {
+			return sdkerrors.Wrap(err, "failed to unmarshal client state bytes into solo machine client state")
+		}
+
+		// update solomachine in store
+		clientStore.Set(host.ClientStateKey(), bz)
+
+		pruneClientConsensusStates(clientStore)
+	}
+
+	return nil
+}
+
+// handlerTendermintMigration asserts that the tendermint client in state can be decoded properly.
+// This ensures the upgrading chain properly registered the tendermint client types on the chain codec.
+func handleTendermintMigration(ctx sdk.Context, store sdk.KVStore, cdc codec.BinaryCodec) error {
+	clients, err := collectClients(ctx, store, exported.Tendermint)
+	if err != nil {
+		return err
+	}
+
+	if len(clients) > 1 {
+		return sdkerrors.Wrap(sdkerrors.ErrLogic, "too many tendermint clients collected")
+	}
+
+	clientID := clients[0]
+
+	clientPrefix := []byte(fmt.Sprintf("%s/%s/", host.KeyClientStorePrefix, clientID))
+	clientStore := prefix.NewStore(store, clientPrefix)
+
+	bz := clientStore.Get(host.ClientStateKey())
+	if bz == nil {
+		return clienttypes.ErrClientNotFound
```

Comment:
should we wrap this error in the same fashion as the other error cases? We could include information about the client state key in the error message.